### ISSUE TITLE
Invalid values in Manhattan Clickhouse table causing API to crash

### ISSUE THREAD

**Issue by JarrodBaker** [opentargets/issues#2670]
There are 1100 studies in the Manhattan table which include invalid entries in the field `top10_genes_coloc_ids`. These cause the API to crash as it attempts to resolve them into genes which cannot be found. This error appears in the API logs as `sangria.execution.deferred.AbsentDeferredValueError: Fetcher has not resolved non-optional ID 'UL'.`. 

## Background

The Manhattan table in Clickhouse is generated by the ETL and uses the outputs of `l2g`, `disease-variant-gene-scored`, `variant-disease-coloc`,  and `variant-disease` as inputs. 

To see a list of studies with invalid entries run query: `select study, top10_genes_coloc_ids from ot.manhattan where has(top10_genes_coloc_ids, NULL)`.

There are two SQL scripts to load the Manhattan table in CH: [manhattan.sql](https://github.com/opentargets/genetics-backend/blob/master/loaders/clickhouse/manhattan.sql), [manhattan_log.sql](https://github.com/opentargets/genetics-backend/blob/master/loaders/clickhouse/manhattan_log.sql). These currently tolerate the NULL values being present which they should not do. 

## Tasks

- [ ] Identify where these nulls are being introduced from. 
- [ ] Ensure that all of the gene IDs listed in the column `top10_genes_coloc_ids` are also in the genes table. 

## Acceptance tests

1. When I run the query `select count(distinct study) from ot.manhattan where has(top10_genes_coloc_ids, NULL)` then 0 rows are returned.

**Comment by ireneisdoomed** [comment]
Trying to find where this could be coming from, I looked for records of one of the offending studies (`FINNGEN_R6_AB1_MUMPS `) where the gene ID was null in the datasets you mention above.

The only one where I found those was in coloc. More specifically, there are 3 rows for this study. I think this is the source of the problem, because if you see the first record, h4 coincides with the assigned score in the top10_genes_coloc_score pairs.

```
-RECORD 0------------------------------------------
 coloc_n_vars              | 2789
 coloc_h0                  | 8.119047467917467E-11
 coloc_h1                  | 6.923574128330692E-5
 coloc_h2                  | 2.995148884136828E-8
 coloc_h3                  | 0.024565975412741796
 coloc_h4                  | 0.9753647588132964 <-------
 left_type                 | gwas
 left_study                | FINNGEN_R6_AB1_MUMPS
 left_chrom                | 3
 left_pos                  | 46589547
 left_ref                  | A
 left_alt                  | AGT
 right_type                | pqtl
 right_study               | SUN2018
 right_bio_feature         | UBERON_0001969
 right_phenotype           | TDGF1_5810_25_3
 right_chrom               | 3
 right_pos                 | 46553679
 right_ref                 | G
 right_alt                 | A
 coloc_h4_h3               | 39.70388891244259
 coloc_log2_h4_h3          | 5.311208418122532
 is_flipped                | true
 right_gene_id             | null
 left_var_right_study_beta | -0.5757
 left_var_right_study_se   | 0.0277
 left_var_right_study_pval | 5.128613839913575E-96
 left_var_right_isCC       | false
-RECORD 1------------------------------------------
 coloc_n_vars              | 2428
 coloc_h0                  | 5.048552266531918E-5
 coloc_h1                  | 0.001540821240299789
 coloc_h2                  | 0.01362251876436524
 coloc_h3                  | 0.4151905104095802
 coloc_h4                  | 0.5695956640630903
 left_type                 | gwas
 left_study                | FINNGEN_R6_AB1_MUMPS
 left_chrom                | 3
 left_pos                  | 46589547
 left_ref                  | A
 left_alt                  | AGT
 right_type                | gwas
 right_study               | GCST005195
 right_bio_feature         | null
 right_phenotype           | null
 right_chrom               | 3
 right_pos                 | 46543532
 right_ref                 | C
 right_alt                 | T
 coloc_h4_h3               | 1.3718898909832773
 coloc_log2_h4_h3          | 0.45616469426066814
 is_flipped                | true
 right_gene_id             | null
 left_var_right_study_beta | null
 left_var_right_study_se   | null
 left_var_right_study_pval | null
 left_var_right_isCC       | null
-RECORD 2------------------------------------------
 coloc_n_vars              | 2428
 coloc_h0                  | 5.048552266531918E-5
 coloc_h1                  | 0.001540821240299789
 coloc_h2                  | 0.01362251876436524
 coloc_h3                  | 0.4151905104095802
 coloc_h4                  | 0.5695956640630903
 left_type                 | gwas
 left_study                | GCST005195
 left_chrom                | 3
 left_pos                  | 46543532
 left_ref                  | C
 left_alt                  | T
 right_type                | gwas
 right_study               | FINNGEN_R6_AB1_MUMPS
 right_bio_feature         | null
 right_phenotype           | null
 right_chrom               | 3
 right_pos                 | 46589547
 right_ref                 | A
 right_alt                 | AGT
 coloc_h4_h3               | 1.3718898909832773
 coloc_log2_h4_h3          | 0.45616469426066814
 is_flipped                | false
 right_gene_id             | null
 left_var_right_study_beta | 0.31543
 left_var_right_study_se   | 0.0752
 left_var_right_study_pval | 2.73E-5
 left_var_right_isCC       | true
```

**Comment by DSuveges** [comment]
@JarrodBaker, @ireneisdoomed  In all these cases the `right_gene_id` is `null`. Indicating this value refers to the right study. If the `right_type` is `gwas` `right_gene_id` field should be null. It should only be mapped to a gene identifier if the `right_type` is some form of qtl (measured gene expression, measured protein level). 

So as far as I understand there's only a problem with the first record, where the `right_phenotype` is a measured protein concentration of `TDGF1`, but because of some reason it is not resolved to [ENSG00000241186](https://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000241186).

**Comment by d0choa** [comment]
Resolving the IDs is done through the `phenotype_id_gene_luts` lookup table

**Comment by DSuveges** [comment]
OK, that's very interesting, when looking at this (`gs://genetics-portal-dev-staging/coloc/220127/coloc_processed_w_betas_fixed.parquet`) coloc dataset, this study looks OK:

```python
(
    spark.read.parquet('coloc_processed_w_betas_fixed.parquet')
    .filter(f.col('right_phenotype') == 'TDGF1_5810_25_3')
    .select('right_study', 'right_type', 'right_phenotype', 'right_gene_id')
    .distinct()
    .show()
)
```

Gives:

```
+-----------+----------+---------------+---------------+
|right_study|right_type|right_phenotype|  right_gene_id|
+-----------+----------+---------------+---------------+
|    SUN2018|      pqtl|TDGF1_5810_25_3|ENSG00000241186|
+-----------+----------+---------------+---------------+
```

Can we assume this problem will be resolved after the release? Or I was not looking at the right dataset. But anyway, I'm quite confident, this is the condition, we need to check:

```python
(
    spark.read.parquet('coloc_processed_w_betas_fixed.parquet')
    .filter(f.col('right_type') != 'gwas')
    .select('right_study', 'right_type', 'right_phenotype', 'right_gene_id')
    .distinct()
    .filter(f.col('right_gene_id').isNull())
    .show()
)
```
This query should return an empty dataframe.

**Comment by JarrodBaker** [comment]
I run @DSuveges filter on the input files used (`gs://genetics-portal-dev-staging/coloc/220408/coloc_processed_w_betas.parquet`) as specified in [Jeremy's spreadsheet](https://docs.google.com/spreadsheets/d/1kXiH9xwpz2eYovm-AOybdaCwy1MiK2aq4bb8L_Ve3zo/edit#gid=0). There are 949 rows in which the `right_type` is `gwas` but the `right_gene_id` is `null`. 

A taster:

```
+-------------+----------+-----------------+-------------+
|  right_study|right_type|  right_phenotype|right_gene_id|
+-------------+----------+-----------------+-------------+
|      SUN2018|      pqtl|IGDCC4_9793_145_3|         null|
|   SUHRE_2017|      pqtl|  ENSG00000156535|         null|
|PIETZNER_2020|      pqtl|  ENSG00000140092|         null|
|      SUN2018|      pqtl|   MFNG_5605_77_3|         null|
|      SUN2018|      pqtl|  MAPK3_2855_49_2|         null|
|      SUN2018|      pqtl|  MANBA_6382_17_3|         null|
|   SUHRE_2017|      pqtl|  ENSG00000154229|         null|
|   SUHRE_2017|      pqtl|  ENSG00000134470|         null|
|      SUN2018|      pqtl| DUSP13_6525_17_3|         null|
|      SUN2018|      pqtl|   CXCL5_2979_8_2|         null|

```
Does the file indicated in the speadsheet also need to be 'fixed'?

**Comment by ireneisdoomed** [comment]
In the 949 rows file, there are 856 unique phenotypes which should be mapped with their gene IDs.
When I merge them with [phenotype_id_gene_luts](https://github.com/opentargets/genetics-colocalisation/tree/master/configs/phenotype_id_gene_luts):
- **506 of them are present in the LUT** and therefore have a mapping to an Ensembl ID.
- **The other 350 are already Ensembl IDs**, so they are not present in the LUT.

So as @DSuveges was pointing out, we don't have any QTLs with a null gene id in the pre ETL coloc dataset. This must be coming from the ETL.

**Comment by DSuveges** [comment]
@ireneisdoomed , I think I might looked the older dataset: `220127 ` vs `220408`. However all the mappings are in this directory: https://github.com/opentargets/genetics-sumstat-data/tree/master/ingest/qtl/pqtl/mappings

These files were added as a single commit, there's not possibilities for version confusion. I think likely a step was skipped. However it needs to be explored. I'm looking at the mappings to make sure we are not missing anythign.

**Comment by JarrodBaker** [comment]
I've manually fixed this in the development database using the following code:
`ALTER TABLE ot.manhattan DELETE WHERE has(top10_genes_coloc_ids, NULL)`
This is just to unblock the FE, and will not persist between machine restarts. It definitely doesn't represent a solution.

**Comment by ireneisdoomed** [comment]
Hi! I've manually corrected the data so that the manual fix doesn't need to happen when ingesting it to CH @JarrodBaker 
The new output is here: `gs://genetics-portal-dev-staging/coloc/220408/coloc_processed_w_betas.parquet`. Same filename, slightly more number of rows, same schema.
The faulty output has been renamed to `coloc_processed_w_betas_obsolete.parquet`.

The only inputs that I needed were the [phenotype to gene](https://github.com/opentargets/genetics-colocalisation/tree/master/configs/phenotype_id_gene_luts) LUTs mentioned above.

Here is the snippet:
```python
coloc_path = 'gs://genetics-portal-dev-staging/coloc/220408/coloc_processed_w_betas.parquet'
phenotype_gene_path = 'gs://ot-team/dochoa/phenotype_id_gene_luts/'

coloc = spark.read.parquet(coloc_path)
phenotype_gene_lut = spark.read.csv(phenotype_gene_path, sep='\t', header=True)

coloc_fix = (
    coloc.join(
        phenotype_gene_lut.withColumnRenamed('gene_id', 'right_gene_id_tmp'),
        coloc['right_phenotype'].eqNullSafe(phenotype_gene_lut['phenotype_id']),
        how='left',
    )
    .drop('phenotype_id')
    .withColumn(
        'right_gene_id',
        F.when(F.col('right_gene_id_tmp').isNull(), F.regexp_extract('right_phenotype', r'(ENSG\d+)', 1)).otherwise(
            F.col('right_gene_id_tmp')
        ),
    )
    .drop('right_gene_id_tmp')
    .distinct()
)
```

And some checks.
```python
coloc.count()
>> 8199256

coloc_fix.count()
>> 8200229

coloc_fix.filter(F.col('right_type') != 'gwas').filter(F.col('right_gene_id').isNull()).count()
>> 0

coloc_fix.filter(F.col('right_type') == 'gwas').filter(F.col('right_gene_id').isNull()).count()
>> 3284288

coloc_fix.schema == coloc.schema
>> True
```

The number of rows have increased because some phenotype IDs (175/41325) map to multiple genes. Do we expect to see this? I have checked one example (`MICB_5102_55_3`) and it seems that the multiple identifiers are referred to the same gene in multiple assemblies. 
```python
+--------------+---------------+
|  phenotype_id|        gene_id|
+--------------+---------------+
|MICB_5102_55_3|ENSG00000231179|
|MICB_5102_55_3|ENSG00000224378|
|MICB_5102_55_3|ENSG00000234218|
|MICB_5102_55_3|ENSG00000204516|
|MICB_5102_55_3|ENSG00000206449|
|MICB_5102_55_3|ENSG00000227772|
|MICB_5102_55_3|ENSG00000238289|
+--------------+---------------+
``

**Comment by ireneisdoomed** [comment]
(Relevant to the above message)

The step of joining the coloc set with the phenotype to gene ID LUT is part of the coloc pipeline (), so the presence of these faulty records means that:
a) The function [load_pheno_to_gene_map](https://github.com/opentargets/genetics-colocalisation/blob/master/6_process_results.py#L232) is not working as expected
a) The coloc pipeline was not entirely run, specifically the module [6_process_results.py](https://github.com/opentargets/genetics-colocalisation/blob/master/6_process_results.py)

Although the dataset has been fixed for now, we still have to investigate where exactly the null genes are coming from.

**Comment by d0choa** [comment]
This will be addressed by the experimental pipeline
